### PR TITLE
Change sites from List to Map

### LINKED ISSUE #4879: long hang on Firefox import
**Did you search for similar issues before submitting this one?** Yes

**Describe the issue you encountered:** Importing Firefox data (history, bookmarks, and cookies) hung the Brave app for a good 30 seconds. Windows reported Brave as "(Not Responding)" for the duration. 

**Expected behavior:** Brave shouldn't hang. Also a progress indicator of some sort would be helpful if it's gonna take that long. 
- Platform (Win7, 8, 10? macOS? Linux distro?): Win10 
- Brave Version: 0.12.4
- Steps to reproduce:
  1. visit settings and import all types data from Firefox
  2. notice brave become unresponsive
  3. check task manager and see Not Responding for about 30 seconds
- Screenshot if needed:
- Any related issues: why no import passwords?

### PR DESCRIPTION
**Pull request by darkdh**
- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).

Test Plan:

fix #4879

Auditors: @bbondy, @bsclifton, @cezaraugusto

Test Plan:
Performance:
1. Import bulk bookmarks (4000+) from other browsers (Don't merge into
                                                      toolbar)
2. The process should finish instantly
3. Delete Import from XXX folder
4. The process should finish instantly

Migration:
1. Make sure session-store-1 contains the sites data before 0.12.10
2. Lauch Brave and Close
3. sites of session-store-1 should change from [] tp {}

### CODE REVIEW

File: js/stores/appStore.js [hunk 1]
```diff
@@ -439,10 +439,15 @@ const handleAppAction = (action) => {
           appState = appState.set('sites', siteUtil.addSite(appState.get('sites'), s, action.tag))
         })
       } else {
-        appState = appState.set('sites', siteUtil.addSite(appState.get('sites'), action.siteDetail, action.tag, action.originalSiteDetail))
+        let sites = appState.get('sites')
+        if (!action.siteDetail.get('folderId') && siteUtil.isFolder(action.siteDetail)) {
```
Comment (darkdh) [inline js/stores/appStore.js]: This is the root cause of last backed out
https://github.com/brave/browser-laptop/commit/766722ad8e33706de2d53b6137e011ce233296a0#diff-23ca389e2bcb77191b5a9c10900eb3a3R443

**Comment by bsclifton** [comment]
Because this can cause issues with migrations, I wanted to propose a new data-structure for us to use in favor of the sites array and ask what everyone thinks?

We could still keep the migration code in there (so that it copies the old values over to the new structure), but instead of: `appState.sites` we could add a new field `appState.siteMap`.  This way, there would be no impact to users if they downgrade. We'd then want to make sure that wiping history removes this field too, of course

**Comment by bbondy** [comment]
This needs to be rebased

**Comment by bbondy** [comment]
Still needs to be rebased, I'll move to 1.0

**Comment by darkdh** [comment]
rebased! but still looking for the cause of multi-select on bookmark manager failure

**Comment by darkdh** [comment]
It's ready for review again

**Comment by posix4e** [comment]
@darkdh  please rebase

**Comment by darkdh** [comment]
@posix4e , rebasing complete!

**Comment by posix4e** [comment]
@darkdh  sorry to be a bother but how was this tested?

**Comment by darkdh** [comment]
@posix4e , I can DM you the testing bookmarks set to do the manual test of test plan 1. You can compare this PR with previous version without it and there is no longer a beach ball when importing.

**Comment by posix4e** [comment]
@darkdh Thanks!

**Comment by luixxiul** [comment]
Also we will want to have a sample file for QA.

**Comment by luixxiul** [comment]
A conflict to be resolved: test/components/urlBarSuggestionsTest.js

**Comment by darkdh** [comment]
rebased. and I will send you the bookmarks set for test

**Comment by cezaraugusto** [comment]
@darkdh when you're done can you provide some bulk bookmark file as your STR suggested? Mine don't reach 300 😭

**Comment by darkdh** [comment]
sent on slack 😄

**Comment by luixxiul** [comment]
Is it OK to make this ready-to-merge? @bsclifton let's do so if you agree :-)

**Comment by bsclifton** [comment]
@luixxiul there are problems with tests currently in the `0.13.1-branch`.  Any help troubleshooting this is appreciated.  We can't accept any more PRs until this is sorted out

cc: @cezaraugusto

**Comment by bbondy** [comment]
@darkdh I think this needs another rebase if you have time between C56 building.

**Comment by darkdh** [comment]
@bbondy, rebased 😄

**Comment by bbondy** [comment]
++

**Comment by luixxiul** [comment]
It's nice to see this finally merged :tada:

**Pull request merged**